src/prefect/infrastructure/kubernetes.py

```diff
@@ -590,20 +591,45 @@ def _watch_job(self, job_name: str) -> int:
         # Wait for job to complete
         # if `job_watch_timeout_seconds` is None, no timeout will be enforced
         self.logger.debug(f"Job {job_name!r}: Starting watch for job completion")
-        watch = kubernetes.watch.Watch()
-        with self.get_batch_client() as batch_client:
-            for event in watch.stream(
-                func=batch_client.list_namespaced_job,
-                field_selector=f"metadata.name={job_name}",
-                namespace=self.namespace,
-                timeout_seconds=self.job_watch_timeout_seconds,
+        start_time = pendulum.now("utc")
+        completed = False
+        while not completed:
+            elapsed = (pendulum.now("utc") - start_time).in_seconds()
+            if (
+                self.job_watch_timeout_seconds is not None
+                and elapsed > self.job_watch_timeout_seconds
             ):
-                if event["object"].status.completion_time:
-                    watch.stop()
-                    break
-            else:
-                self.logger.error(f"Job {job_name!r}: Job did not complete.")
+                self.logger.error(f"Job {job_name!r}: Job timed out.")
                 return -1
+            try:
+                watch = kubernetes.watch.Watch()
+                with self.get_batch_client() as batch_client:
+                    remaining_timeout = (
+                        (  # subtract previous watch time
+                            self.job_watch_timeout_seconds - elapsed
+                        )
+                        if self.job_watch_timeout_seconds
+                        else None
+                    )
+
+                    for event in watch.stream(
+                        func=batch_client.list_namespaced_job,
+                        field_selector=f"metadata.name={job_name}",
+                        namespace=self.namespace,
+                        timeout_seconds=remaining_timeout,
+                    ):
+                        if event["object"].status.completion_time:
+                            watch.stop()
+                            completed = True
+                            break  # don't enter else, just end watch
+                    else:
+                        self.logger.error(f"Job {job_name!r}: Job did not complete.")
+                        return -1
```

Comment:
We were entering this case before, so I think we need to check start the watch again if the job is still in a running state.

Reply:
We may be able to just remove this entirely since it's been replaced with the logic on L602?

Reply:
ah yeah, I guess was thinking the for loop could still somehow complete without seeing the completion time, in which case it seemed like something could be wrong